### DISCUSSION TITLE
Questions on River protocol architecture

### DISCUSSION THREAD

**Discussion by AnthonyAkentiev** [river-build/river#324] [General]
Hey. Tony is here.
Very interesting project and implementation!

I have a couple of questions:
1) It seems like nodes/operators specify the exact location of each Stream in the Stream Registry ->  https://github.com/river-build/river/blob/main/contracts/src/river/registry/facets/stream/StreamRegistry.sol#L23 

So there is no means for a user to "move" or "migrate" from one node to another if it censors him?

2) How do I retrieve a Stream ID by Space ID or Space URL?
It looks like there is no need for some special table or mapping? And Space ID can be converted to the Stream ID deterministically. 
So there is no special "Space Registry"?

Thank you!

**Comment by AnthonyAkentiev** [comment]
@clemire @tak-hntlabs @jterzis Maybe some of you guys can answer my questions, please? Will be really grateful.

**Reply by tak-hntlabs** [reply]
Excellent questions.
RE: about the registry and nodes
The registry helps with several use cases. One of which is to bootstrap the initial owner of each stream. Right now, all the nodes are also stream forwarders.  If a node is unaware of a stream, it will make a sync request to the node owner to sync that stream and will then forward it to the client. Thus, a client can connect to any node, and be able to get all the streams that it is interested in regardless of whether that connected node is the owner of a stream or not. We will be adding stream replication and archiving to beef up the resilience of data availability. Work in progress.

RE: a malicious node censoring users. We are working towards a protocol implementation which will reward honest nodes, and punish dishonest nodes. Our vision is to have a protocol + implementation such that nodes in the network are kept honest, and any dishonest nodes are detected and kicked out. Not there yet. Now it is still in the high trust stage.

RE: retrieving the streamID, spaceID, spaceURL from a client. Roughly, the logged in user would retrieve its own user stream. From its own user stream, it can parse the stream events to retrieve its memberships in other streams (towns, channels). Then it can explicitly ask the node for those streams. If you are using the towns app, then the URL follows a pattern like "https://app.towns.com/t/the_town_stream_id/......" etc. 
All of these is in our internal SDK. Doing this without our SDK is definitely a lot of work. We understand that, and are working towards releasing an SDK. No ETA.

**Reply by AnthonyAkentiev** [reply]
@tak-hntlabs Thank you very much for the answer! It clarified some things.

Still I didn't get how others can find a Stream in this scenario:

1. A malicious node A writes that stream ABC's location is node B -> Stream Registry
2. Node B has no stream ABC data
3. Other guys ask random node/node B to retrieve stream ABC contents
4. What will happen? You will slash Node A? But how do we find the Stream ABC's content even if node A got slashed or removed from the list?

**Reply by tak-hntlabs** [reply]
> @tak-hntlabs Thank you very much for the answer! It clarified some things.
> 
> Still I didn't get how others can find a Stream in this scenario:
> 
> 1. A malicious node A writes that stream ABC's location is node B -> Stream Registry
> 2. Node B has no stream ABC data
> 3. Other guys ask random node/node B to retrieve stream ABC contents
> 4. What will happen? You will slash Node A? But how do we find the Stream ABC's content even if node A got slashed or removed from the list?

The Registry smart contract isn't openly writeable by any random nodes. A malicious node can't simply show up on the network and write fake data into the registry. There is a delegation / governance flow that I am not familiar with. Only highly trusted nodes as assigned by the governance process is allowed to update the registry. Our intent is to make this fully automated, and self-running. Not there yet. 

Adding our protocol lead @sergekh2 to describe in more details, and correct any of my misunderstandings.

**Reply by AnthonyAkentiev** [reply]
@tak-hntlabs Thank you!

**Reply by quietBlockchain** [reply]
Thanks for the detailed responses!

It looks like most of this logic is currently accessible only via the internal SDK.  
Do you have any plans to expose stream-to-node resolution or registry insights through a public API or a simple explorer interface?

Such a solution would be very useful for developers who want to build community tools without relying on the SDK.

**Reply by clemire** [reply]
cc @Crisvond-hnt

**Reply by clemire** [reply]
@quietBlockchain just an FYI, we've recently changed our node allocation algorithm to favor nodes with smaller numbers of streams assigned and also to favor distributing node allocation across operators. The main goal here was maximizing availability of replicated streams and evenly distributing the storage/request load among nodes for the sake of fairness.

None of this logic is directly exposed through the API; you can only invoke the allocation by creating a new stream, and the assignment of nodes updates dynamically, via pseudo-randomness. I can link you the specific go file, if you'd like. Cris can confirm what APIs/insights we plan to expose.

**Reply by quietBlockchain** [reply]
Thanks for the update @clemire! That makes sense and I appreciate the transparency.

I’d definitely be interested in checking out the Go file you mentioned that would help better understand how node allocation works under the hood.

Also curious: if node updates are pseudo-random and dynamic, will there eventually be a way for third-party devs to fetch those mappings via a read-only endpoint or light API?

**Reply by clemire** [reply]
@quietBlockchain take a look [here](https://github.com/towns-protocol/towns/blob/a6e5332458e1a37caf2951f74432bc159ed4442a/core/node/nodes/streamplacement/distribution.go#L159).

**Reply by clemire** [reply]
@quietBlockchain I'm actually a bit unsure of what information you're hoping may become available via API. Would you mind sharing the context of what you'd like to do?

**Reply by quietBlockchain** [reply]
@clemire Thanks for the follow-up, really appreciate it!

Our initial idea was to build an open analysis tool not to interact with the stream placement logic directly, but to analyze on-chain StreamUpdated events and show how the current node allocation behaves in practice.

Basically, we’d track:

How many streams are assigned to each node

How balanced the stream distribution is across operators

Whether pseudo-randomness fairly reflects the replication logic

Even without direct API access, event logs and contract calls provide enough data to build a public dashboard or explorer that gives visibility into the current state of node allocation.

If helpful, we can start working on a prototype and share early results with the community!

**Reply by clemire** [reply]
@quietBlockchain apologies for the delay in my response! You are more than welcome to build such a tool, and to analyze any particular interesting trait of the network - it is, after all, public! And we are really excited to see your engagement. :)

If you find any issues with our replication strategy, we'd also be happy to triage it against other growth challenges on the network, and I personally appreciate the quality feedback.

Let me know what you decide, and happy building! <3
